File: reservas.py
# Matriz doblemente anidada se hace esta ya que necesita guardar 4 datos por reserva
reservas = []
# Función para agresar una nueva reserva
def ingresar_reserva():
    cod_est = input("Código del estudiante: ")
    nom_est = input("Nombre del estudiante: ")
    cod_lib = input("Código del libro: ")
    nom_lib = input("Nombre del libro: ")

    reservas.append([cod_est, nom_est, cod_lib, nom_lib])
    print("✅ Reserva agregada correctamente.\n")
# Función para buscar una reserva
def buscar_reserva():
    print("Buscar por:")
    print("1. Código de estudiante")
    print("2. Código de libro")
    opcion = input("Seleccione opción: ")

    if opcion == "1":
        cod_est = input("Ingrese código del estudiante: ")
        encontrados = [r for r in reservas if r[0] == cod_est]
    elif opcion == "2":
        cod_lib = input("Ingrese código del libro: ")
        encontrados = [r for r in reservas if r[2] == cod_lib]
    else:
        print("⚠️ Opción inválida.\n")
        return

    if encontrados:
        print("🔎 Resultados:")
        for r in encontrados:
            print(f"Estudiante: {r[1]} ({r[0]}) - Libro: {r[3]} ({r[2]})")
    else:
        print("⚠️ No se encontraron registros.")
    print()
# Función para eliminar una reserva
def eliminar_reserva():
    cod_est = input("Código del estudiante: ")
    cod_lib = input("Código del libro: ")

    indice = -1
    for i, r in enumerate(reservas):
        if r[0] == cod_est and r[2] == cod_lib:
            indice = i
            break

    if indice != -1:
        reservas.pop(indice)
        print("🗑️ Reserva eliminada correctamente.\n")
    else:
        print("⚠️ No se encontró esa reserva.\n")
# Funcion para mostrar todas las reservas de ser necesario
def mostrar_reservas():
    if not reservas:
        print("📂 No hay reservas registradas.\n")
    else:
        print("📚 Lista de reservas:")
        for r in reservas:
            print(f"Estudiante: {r[1]} ({r[0]}) - Libro: {r[3]} ({r[2]})")
        print()
# Menú interactivo para el usuario y la ejecución del programa
def menu():
    opcion = ""
    while opcion != "5":
        print("====== MENÚ RESERVAS ======")
        print("1. Ingresar reserva")
        print("2. Buscar reserva")
        print("3. Eliminar reserva")
        print("4. Mostrar todas las reservas")
        print("5. Salir")
        opcion = input("Seleccione una opción: ")

        if opcion == "1":
            ingresar_reserva()
        elif opcion == "2":
            buscar_reserva()
        elif opcion == "3":
            eliminar_reserva()
        elif opcion == "4":
            mostrar_reservas()
        elif opcion == "5":
            print("👋 Saliendo del sistema...")
        else:
            print("⚠️ Opción inválida.\n")
# Ejecutar el programa
menu()
import time
def ingresar_reserva(cod_est, nom_est, cod_lib, nom_lib):
    reservas.append([cod_est, nom_est, cod_lib, nom_lib])

def buscar_reserva(cod_est=None, cod_lib=None):
    if cod_est:
        return [r for r in reservas if r[0] == cod_est]
    elif cod_lib:
        return [r for r in reservas if r[2] == cod_lib]
    return []

def eliminar_reserva(cod_est, cod_lib):
    for i, r in enumerate(reservas):
        if r[0] == cod_est and r[2] == cod_lib:
            reservas.pop(i)
            return True
    return False
def probar_escenario(n):
    global reservas
    reservas = []  # Reiniciar lista

    print(f"\n--- Escenario con {n} reservas ---")

    # Medir tiempo de inserción
    inicio = time.time()
    for i in range(n):
        ingresar_reserva(f"E{i}", f"Estudiante{i}", f"L{i}", f"Libro{i}")
    fin = time.time()
    print(f"Inserción de {n} reservas: {fin - inicio:.6f} segundos")

    # Medir tiempo de búsqueda (último registro = peor caso)
    inicio = time.time()
    buscar_reserva(cod_est=f"E{n-1}")
    fin = time.time()
    print(f"Búsqueda de un estudiante: {fin - inicio:.6f} segundos")

    # Medir tiempo de eliminación (último registro = peor caso)
    inicio = time.time()
    eliminar_reserva(f"E{n-1}", f"L{n-1}")
    fin = time.time()
    print(f"Eliminación de una reserva: {fin - inicio:.6f} segundos")


# Se ejecuta esta prueba para evaluar el rendimiento del sistema

for cantidad in [10, 100, 1000]:
    probar_escenario(cantidad) 

File: reservas.js
const prompt = require("prompt-sync")(); // Para poder ingresar los datos por cosola 

let reservas = [];
// Función para ingresar una nueva reserva
function ingresarReserva() {
  const codigoEstudiante = prompt("Código del estudiante: ");
  const nombreEstudiante = prompt("Nombre del estudiante: ");
  const codigoLibro = prompt("Código del libro: ");
  const nombreLibro = prompt("Nombre del libro: ");

  reservas.push([codigoEstudiante, nombreEstudiante, codigoLibro, nombreLibro]);
  console.log("✅ Reserva agregada correctamente.\n");
}
// Función para buscar por código de estudiante o código de libro
function buscarReserva() {
  const opcion = prompt("Buscar por Estudiante (1) o Buscar por Libro (2):  ");

  if (opcion === "1") {
    const codigoEstudiante = prompt("Ingrese el código del estudiante: ");
    const encontrados = reservas.filter(r => r[0] === codigoEstudiante);

    if (encontrados.length > 0) {
      console.log("🔎 Resultados:");
      encontrados.forEach(r =>
        console.log(`Estudiante: ${r[1]} - Libro: ${r[3]} (${r[2]})`)
      );
    } else {
      console.log("⚠️ No se encontró ninguna reserva para ese estudiante.");
    }
  } else if (opcion === "2") {
    const codigoLibro = prompt("Ingrese el código del libro: ");
    const encontrados = reservas.filter(r => r[2] === codigoLibro);

    if (encontrados.length > 0) {
      console.log("🔎 Resultados:");
      encontrados.forEach(r =>
        console.log(`Libro: ${r[3]} - Estudiante: ${r[1]} (${r[0]})`)
      );
    } else {
      console.log("⚠️ No se encontró ninguna reserva para ese libro.");
    }
  } else {
    console.log("⚠️ Opción inválida.");
  }
  console.log();
}
// Función para eliminar una reserva
function eliminarReserva() {
  const codigoEstudiante = prompt("Código del estudiante: ");
  const codigoLibro = prompt("Código del libro: ");

  const indice = reservas.findIndex(
    r => r[0] === codigoEstudiante && r[2] === codigoLibro
  );

  if (indice !== -1) {
    reservas.splice(indice, 1);
    console.log("🗑️ Reserva eliminada correctamente.\n");
  } else {
    console.log("⚠️ No se encontró ninguna reserva con esos datos.\n");
  }
}

// M
// Función para mostrar todas las reservas de ser necesario
function mostrarReservas() {
  if (reservas.length === 0) {
    console.log("📂 No hay reservas registradas.\n");
  } else {
    console.log("📚 Lista de reservas:");
    reservas.forEach(r =>
      console.log(
        `Estudiante: ${r[1]} (${r[0]}) - Libro: ${r[3]} (${r[2]})`
      )
    );
    console.log();
  }
}
// Menú interactivo

function menu() {
  let opcion = "";
  while (opcion !== "5") {
    console.log("====== MENÚ RESERVAS ======");
    console.log("1. Ingresar reserva");
    console.log("2. Buscar reserva");
    console.log("3. Eliminar reserva");
    console.log("4. Mostrar todas las reservas");
    console.log("5. Salir");
    opcion = prompt("Seleccione una opción: ");

    switch (opcion) {
      case "1":
        ingresarReserva();
        break;
      case "2":
        buscarReserva();
        break;
      case "3":
        eliminarReserva();
        break;
      case "4":
        mostrarReservas();
        break;
      case "5":
        console.log("👋 Saliendo del sistema...");
        break;
      default:
        console.log("⚠️ Opción inválida.\n");
    }
  }
}
// Ejecutar el menú
menu();